kong/tools/stream_api.lua

```diff
@@ -5,114 +5,355 @@
-local st_format  = string.format
-local table_concat = table.concat
-local assert     = assert
+local CLOSE = 444
+local OK = 0
 
-local MAX_DATA_LEN = 8000
-local PREFIX = ngx.config.prefix()
+
+local PACK_F = "=CI"
+
+-- unsigned char length
+local MAX_KEY_LEN = 2^8 - 1
+
+-- since the length is represented by an unsigned int we could theoretically
+-- go up to 2^32, but that seems way beyond the amount of data that we should
+-- expect to see exchanged over this interface
+local MAX_DATA_LEN = 2^22 - 1
+
+local HEADER_LEN = #st_pack(PACK_F, MAX_KEY_LEN, MAX_DATA_LEN)
+
+local SOCKET_PATH = "unix:" .. ngx.config.prefix() .. "/stream_rpc.sock"
 
 local stream_api = {}
 
 local _handlers  = {}
 
 
+-- # RPC format
+--
+-- RPC messages have a header and a body that are slightly different between
+-- request and response.
+--
+-- ## requests
+--
+-- Requests have two components:
+--  * key (string)
+--  * payload (string)
+--
+-- The request header is made up of:
+--
+-- |   key len     | payload len  |
+-- +---------------+--------------+
+-- | unsigned char | unsigned int |
+--
+-- The header is followed by the request body, which is simply the request key
+-- followed by the request payload (no separator).
+--
+-- ## responses
+--
+-- Responses have two components:
+--   * status (integer)
+--   * payload (string)
+--
+-- Responses have the same header length as requests, but with different
+-- meanings implied by each field:
+--
+-- |   status      | body/payload len  |
+-- +---------------+-------------------+
+-- | unsigned char |   unsigned int    |
+--
+-- The response header is followed by the response body, which is equal to the
+-- body/payload size in the header.
+
+
+--- Compose and send a stream API message.
+--
+-- Returns truth-y on success or `nil`, and an error string on failure.
+--
+-- @tparam  tcpsock       sock
+-- @tparam  string|number key_or_status
+-- @tparam  string        data
+-- @treturn boolean       ok
+-- @treturn string|nil    error
+local function send(sock, key_or_status, data)
+  local key
+  local key_len
+
+  local typ = type(key_or_status)
+
+  if typ == "number" then
+    -- we're sending a response, so the (numerical) status is simply encoded as
+    -- part of the header
+    key = ""
+    key_len = key_or_status
+
+  elseif typ == "string" then
+    -- we're sending a request, so the key length is included in the header,
+    -- while the key itself is part of the body
+    key = key_or_status
+    key_len = #key_or_status
+
+    if key_len == 0 then
+      return nil, "empty key"
+    end
+
+  else
+    return nil, "invalid type for key/status: " .. typ
+  end
+
+  if key_len > MAX_KEY_LEN then
+    return nil, "max key/status size exceeded"
+  end
+
+  local data_len = #data
+  if data_len > MAX_DATA_LEN then
+    return nil, "max data size exceeded"
+  end
+
+  local header = st_pack(PACK_F, key_len, data_len)
+  local msg = header .. key .. data
+
+  return sock:send(msg)
```

Comment:
This might be faster and/or less memory intensive.  or maybe not.  creating temporary strings is usually bad idea; creating non-expanding tables is usually faster but in some cases the JIT determines that the string never escapes a tight trace so the creation could be sunk.  The only way to know is to test!

```suggestion
  return sock:send({header, key, data})
```

Reply:
Right, so I did some testing around this earlier on. There wasn't a clear winner in terms of RPS (at least for sensible payload sizes), but I didn't measure anything else, so maybe memory usage would be a different story. My gut tells me that there's probably a threshold of payload size where the temp table approach starts to consistently win out over string concatenation, but I'd need to go a little deeper to find out for sure.